healthcheck/healthcheck.go

```diff
@@ -0,0 +1,115 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package healthcheck
+
+import (
+	"errors"
+	"io"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/internal"
+	"google.golang.org/grpc/internal/backoff"
+	"google.golang.org/grpc/status"
+)
+
+func init() {
+	internal.HealthCheckFunc = newClientHealthCheck
+}
+
+const maxDelay time.Duration = 5 * time.Second
+
+func newClientHealthCheck(ctx context.Context, newStream func() (interface{}, error), reportHealth func(bool), service string) error {
+	retryCnt := 0
+	needsBackoff := false
+	bo := backoff.Exponential{MaxDelay: maxDelay}
+
+retryConnection:
+	for {
+		// Backs off if the connection has failed in some way without receiving a message in the previous retry.
+		if needsBackoff {
+			timer := time.NewTimer(bo.Backoff(retryCnt))
+			select {
+			case <-timer.C:
+			case <-ctx.Done():
+				timer.Stop()
+				return nil
+			}
+			retryCnt++
+		}
+
+		// Assumes that the connection will fail without receiving a message and we will need a backoff before the next retry.
+		// If everything goes alright and we receive a message before connection fails, then this is set to false below.
+		needsBackoff = true
+
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+		rawS, err := newStream()
+		if err != nil {
+			continue retryConnection
+		}
+
+		s, ok := rawS.(grpc.ClientStream)
+		if !ok {
+			// exit the health check function
+			return errors.New("type assertion to grpc.ClientStream failed")
```

Comment:
I think this one is logged here: https://github.com/grpc/grpc-go/pull/2387/files#diff-e1550a73f5d25064c8b586ec68d81a64R1291

cc @lyuxuan

Reply:
It shouldn't be an "error" level log when the server has health checks disabled.  Probably just "info" or _maybe_ "warning".

This one, on the other hand, should be logged as an error, since it's an internal invariant that should never be broken.  Let's log it separately here.

As for what we log, please include the type that we did see (even though this is literally impossible so it will never happen):

```go
grpclog.Errorf("newStream returned %v (type %T); want grpc.ClientStream", rawS, rawS)
```

Reply:
Actually after discussing offline, I agree with the other logs being "error": since the user has gone through great troubles to turn on health checking (importing this package, setting up service config, and using a balancer that supports it), then if it doesn't work, that could indicate "an error in their usage of grpc" (one criterion for the error log level).

Please set the channel as healthy when this unexpected event happens, too, though.